### PR TITLE
Implement json functions

### PR DESCRIPTION
**Pull request by murphyatwork**
## What type of PR is this：
- [ ] bug
- [x] feature
- [ ] enhancement
- [ ] others

## Summary
- Supported json functions: `parse_json`, `json_query`, `json_exists`, `json_object`, `json_array`

### CODE REVIEW

File: be/src/exprs/vectorized/json_functions.cpp [hunk 1]
```diff
@@ -337,27 +353,255 @@ ColumnPtr JsonFunctions::get_json_string(FunctionContext* context, const Columns
...
+        result.append(&json);
+    }
+    return result.build(ColumnHelper::is_all_const(columns));
+}
+
+ColumnPtr JsonFunctions::json_object(FunctionContext* context, const Columns& columns) {
+    namespace vpack = arangodb::velocypack;
+
+    // empty arguments create an empty object
+    if (columns.size() == 0) {
```
Comment (kangkaisen) [inline be/src/exprs/vectorized/json_functions.cpp:547]: This process way seems has an issue. please test it with real data set. The result column size shouldn't always 1.

File: be/src/exprs/vectorized/json_functions.h [hunk 2]
```diff
@@ -115,24 +126,45 @@ class JsonFunctions {
...
+
+    /**
+     * Build json array from json values
      * @param: [json_object, ...]
      * @paramType: [JsonColumn, ...]
      * @return: JsonColumn
      */
     DEFINE_VECTORIZED_FN(json_array);
 
+    static Status native_json_path_prepare(starrocks_udf::FunctionContext* context,
```
Comment (kangkaisen) [inline be/src/exprs/vectorized/json_functions.h:151]: Why  the method name need to start with `native`?

File: be/src/column/column.h [hunk 3]
```diff
@@ -72,6 +72,8 @@ class Column {
 
     virtual bool is_object() const { return false; }
 
+    virtual bool is_json() const { return false; }
```
Comment (kangkaisen) [inline be/src/column/column.h:75]: Do we really need  this method?

File: be/src/exprs/vectorized/json_functions.h (hunk 2 above)
Comment (mofeiatwork) [reply be/src/exprs/vectorized/json_functions.h:151]: There's already a function named `json_path_prepare`, for string-based JSON implementation. So the new JSON type is called `Native Json` to differentiate.

File: be/src/column/column.h (hunk 3 above)
Comment (mofeiatwork) [reply be/src/column/column.h:75]: Temporally used, could be removed.

File: be/src/exprs/vectorized/json_functions.cpp (hunk 1 above)
Comment (mofeiatwork) [reply be/src/exprs/vectorized/json_functions.cpp:547]: Fixed. 

Actually this function could not be matched with empty arguments, and throws an error `No matching function with signature: json_array`.

So I another function signature `json_object()` with empty arguments, and another function implementations.

**Review by kangkaisen — DISMISSED** [review]

**Comment by kangkaisen** [comment]
The overall implementation LGTM, but the function detail still need other people review.

File: be/src/exprs/vectorized/jsonpath.h [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+    int left, right;
+
+    ArraySelectorSlice(int left, int right) : left(left), right(right) { type = SLICE; }
+
+    static bool match(const std::string& input);
+
+    void iterate(vpack::Slice array_slice, std::function<void(vpack::Slice)> callback) override;
+};
+
+// JsonPath implement that suuport array building
```
Comment (Pslydhh) [inline be/src/exprs/vectorized/jsonpath.h:73]:
```suggestion
// JsonPath implement that support array building
```

File: be/src/exprs/vectorized/json_functions.cpp [hunk 5]
```diff
@@ -8,29 +8,45 @@
...
+#include "udf/udf.h"
+#include "util/json.h"
+#include "velocypack/vpack.h"
 
 namespace starrocks::vectorized {
 
 // static const re2::RE2 JSON_PATTERN("^([a-zA-Z0-9_\\-\\:\\s#\\|\\.]*)(?:\\[([0-9]+)\\])?");
 // json path cannot contains: ", [, ]
-static const re2::RE2 JSON_PATTERN(R"(^([^\"\[\]]*)(?:\[([0-9]+|\*)\])?)");
+const re2::RE2 JSON_PATTERN(R"(^([^\"\[\]]*)(?:\[([0-9\:\*]+)\])?)");
```
Comment (stdpain) [inline be/src/exprs/vectorized/json_functions.cpp:28]: what the meaning for `$a.b[::]`?

Comment (mofeiatwork) [reply be/src/exprs/vectorized/json_functions.cpp:28]: The modification is for `$.a.b[1:3]`, known as array slicing.

Comment (stdpain) [reply be/src/exprs/vectorized/json_functions.cpp:28]: It looks like this regular will match an unintended case. such as `$.a.b[::]`, `$.a.b[:*]`

File: be/src/exprs/vectorized/json_functions.cpp [hunk 6]
```diff
@@ -150,14 +166,14 @@ bool JsonFunctions::extract_from_object(simdjson::ondemand::object& obj, const s
     return ok;
 }
 
-void JsonFunctions::parse_json_paths(const std::string& path_string, std::vector<JsonPath>* parsed_paths) {
+void JsonFunctions::parse_json_paths(const std::string& path_string, std::vector<SimpleJsonPath>* parsed_paths) {
     // split path by ".", and escape quota by "\"
     // eg:
     //    '$.text#abc.xyz'  ->  [$, text#abc, xyz]
     //    '$."text.abc".xyz'  ->  [$, text.abc, xyz]
     //    '$."text.abc"[1].xyz'  ->  [$, text.abc[1], xyz]
```
Comment (ZiheLiu) [inline be/src/exprs/vectorized/json_functions.cpp:174]:
```suggestion
    //    '$."text.abc[1]".xyz'  ->  [$, text.abc[1], xyz]
```

File: be/src/exprs/vectorized/json_functions.cpp [hunk 7]
```diff
@@ -8,29 +8,45 @@
...
+        if (i == 0) {
+            if (current != "$") {
+                parsed_paths->emplace_back("", -1, true);
+            } else {
+                parsed_paths->emplace_back("$", -1, true);
+                continue;
+            }
+        }
+
         if (UNLIKELY(!RE2::FullMatch(path_exprs[i], JSON_PATTERN, &col, &index))) {
```
Comment (ZiheLiu) [inline be/src/exprs/vectorized/json_functions.cpp:50]: `index` may be `"1:3"`, why does the `else` branch always consider it as an integer? Or maybe we should use different `JSON_PATTERN` between here and `JsonPathPiece::parse`?

File: be/src/exprs/vectorized/json_functions.cpp (hunk 6 above)
Comment (mofeiatwork) [reply be/src/exprs/vectorized/json_functions.cpp:174]: Emm, looks like you have some misunderstanding over it.

The `"` in jsonpath is used to quote json field name, like `text.abc`. So `'$."text.abc"[1].xyz'` represents json object `{"text.abc": [{"xyz": 1}]}`. I think the former is the right syntax.

File: be/src/exprs/vectorized/json_functions.cpp (hunk 7 above)
Comment (mofeiatwork) [reply be/src/exprs/vectorized/json_functions.cpp:50]: Well, this function only deal with SimpleJsonPath, which does not support `arr[1:3]` syntax.

The `SimpleJsonPath` is previous implementation, and the `JsonPath` is yet another implementation, covers more complex jsonpath syntax like `arr[1:3]`, `arr[1].xxx`, `arr[*].xxx`.

File: be/src/exprs/vectorized/json_functions.cpp (hunk 5 above)
Comment (mofeiatwork) [reply be/src/exprs/vectorized/json_functions.cpp:28]: Well, this regex does not validate the unintended case. But `ArraySelector::parse` will do more validate work, these illegal jsonpath will get a `Invalid json path` error.

File: be/src/exprs/vectorized/json_functions.cpp (hunk 7 above)
Comment (ZiheLiu) [reply be/src/exprs/vectorized/json_functions.cpp:50]: @mofeiatwork This function only deals with `SimpleJsonPath`, but also uses `JSON_PATTERN(R"(^([^\"\[\]]*)(?:\[([0-9\:\*]+)\])?)")`, which also matches the string like `$.a.b[1:3]`. Therefore, should we use a different `JSON_PATTERN` here?

Comment (mofeiatwork) [reply be/src/exprs/vectorized/json_functions.cpp:50]: Good idea. I will fix it soon.

File: be/src/exprs/vectorized/json_functions.cpp (hunk 6 above)
Comment (ZiheLiu) [reply be/src/exprs/vectorized/json_functions.cpp:174]: @mofeiatwork Will `'$."text.abc"[1].xyz'` be tokennized to `["$", "text.abc", "[1]", "xyz"]`, not `[$, text.abc[1], xyz]` if as it does?

Comment (mofeiatwork) [reply be/src/exprs/vectorized/json_functions.cpp:174]: `boost::escapted_list_separator` accept three parameters, `Escape character`, `Comma character`, `Quote character`. So above two representation `$."text.abc"[1].xyz` and `$."text.abc[1]".xyz` are same.


```c++
    explicit escaped_list_separator(Char  e = '\\',
                                    Char c = ',',Char  q = '\"')
      : escape_(1,e), c_(1,c), quote_(1,q), last_(false) { }
```

File: be/src/exprs/vectorized/json_functions.cpp (hunk 5 above)
Comment (mofeiatwork) [reply be/src/exprs/vectorized/json_functions.cpp]: As @ZiheLiu suggested, I have separated it into two regex to validate, as `SIMPLE_JSONPATH_PATTERN` and `JSONPATH_PATTERN`.

**Review by kangkaisen — APPROVED** [review]

**Review by stdpain — APPROVED** [review]

**Review by ZiheLiu — APPROVED** [review]

**Pull request merged**